macaroons/constraints.go

```diff
@@ -0,0 +1,86 @@
+package macaroons
+
+import (
+	"fmt"
+	"net"
+	"time"
+
+	"gopkg.in/macaroon-bakery.v1/bakery/checkers"
+	macaroon "gopkg.in/macaroon.v1"
+)
+
+// Constraint type adds a layer of indirection over macaroon caveats and
+// checkers.
+type Constraint func(*macaroon.Macaroon)
+
+// AddConstraints returns new derived macaroon by applying every passed
+// constraint and tightening its restrictions.
+func AddConstraints(mac *macaroon.Macaroon, cs ...Constraint) *macaroon.Macaroon {
+	newMac := mac.Clone()
+	for _, constraint := range cs {
+		constraint(newMac)
+	}
+	return newMac
+}
+
+// Each *Constraint function is a functional option, which takes a pointer
+// to the macaroon and adds another restriction to it. For each *Constraint,
+// the corresponding *Checker is provided.
+
+// AllowConstraint restricts allowed operations set to the ones
+// passed to it.
+func AllowConstraint(ops ...string) func(*macaroon.Macaroon) {
+	return func(mac *macaroon.Macaroon) {
+		caveat := checkers.AllowCaveat(ops...)
+		mac.AddFirstPartyCaveat(caveat.Condition)
+	}
+}
+
+// AllowChecker wraps default checkers.OperationChecker.
+func AllowChecker(method string) checkers.Checker {
+	return checkers.OperationChecker(method)
+}
+
+// TimeoutConstraint restricts the lifetime of the macaroon
+// to the amount of seconds given.
+func TimeoutConstraint(seconds int64) func(*macaroon.Macaroon) {
+	return func(mac *macaroon.Macaroon) {
+		macaroonTimeout := time.Duration(seconds)
+		requestTimeout := time.Now().Add(time.Second * macaroonTimeout)
+		caveat := checkers.TimeBeforeCaveat(requestTimeout)
+		mac.AddFirstPartyCaveat(caveat.Condition)
+	}
+}
+
+// TimeoutChecker wraps default checkers.TimeBefore checker.
+func TimeoutChecker() checkers.Checker {
+	return checkers.TimeBefore
+}
+
+// IPLockConstraint locks macaroon to a specific IP address.
+// If address is an empty string, this constraint does nothing to
+// accommodate default value's desired behavior.
+func IPLockConstraint(ipAddr string) func(*macaroon.Macaroon) {
+	return func(mac *macaroon.Macaroon) {
+		if ipAddr != "" {
+			macaroonIPAddr := net.ParseIP(ipAddr)
```

Comment:
This checks whether `ipAddr`, received from `-macaroonip` option, is not a default `""` value, and if it isn't, adds another caveat.

Reply:
Looks good, thanks for pointing this out. One more thing I noticed - if `net.ParseIP` returns nil because it can't parse the IP, you get a caveat like `error bad IP address []`, and there's no error check nor any way to return that error. You can try this with, e.g. `lncli --macaroonip 127.0.0.0/8 getinfo` (using a CIDR, which isn't supported).

It would probably be better to get an error client-side before sending a malformed macaroon. I'd do this by changing the definition of `Constraint` to be a `func (*macaroon.Macaroon) error` so that it could return an error in case the caveat parameters are incorrect some way. Then, when applying the options in `ApplyConstraints`, check for the error and return the error as needed.

Reply:
This is a great suggestion. Current impl. is broken without it. Done in the tip of the branch. Strange, but I can't see my last two commits in this PR.